File: code/Scanner.ts
import {AudioMetaData} from "./AudioMetaData";
import fs = require('fs');
import {Global} from "./Global";
import {Database} from "./Database";
import child_process = require('child_process');

export class Scanner{
    db: Database;
    baseDir: string;
    scanlist: string[] = [];

    constructor(db: Database, baseDir: string){
        this.db = db;
        this.baseDir = baseDir;
    }

    public scan(){
        this.scanDirectory(this.baseDir);
    }

    public rescan(){
        
    }

    public validate(){

    }

    scanDirectory(dirname: string){
        var jobid = Global.JobControl.createJob("Scanner","Scan Directory", "Scanning " + dirname);
        var dirs = fs.readdirSync(dirname);
        var jobs = [];
        dirs.forEach(dir => {
            if(!dir.startsWith(".")){
                var stat = fs.statSync(dirname + "/" + dir);
                if(stat.isDirectory()){
                    this.scanDirectory(dirname + "/" + dir);
                }else if(stat.isFile() && dir.endsWith(".mp3")){
                    jobs.push(this.scanFile(dirname + "/" + dir));
                }
            }
        });
        Promise.all(jobs).then(() => Global.JobControl.finishJob(jobid));
    }

    handleScan():Promise<void>{
        return new Promise<void>((resolve, reject) => {
            if(this.scanlist.length > 0){
                var toScan = this.scanlist.pop();
                var stat = fs.statSync(toScan);
                if(stat.isDirectory()){
                    this.scanDirectory(toScan);
                }
            }
        });
        
    }

    public scanFile(filename: string):Promise<void>{
        return new Promise<void>((resolve, reject) => {
            this.db.containsFile(filename).then(contains => {
                if(!contains){
                    //var jobid = Global.JobControl.createJob("Scanner","Scan File", "Scanning " + filename);
                    var options = {
                        "stdio" : [null, "pipe", null]
                    };
                    var stdout = child_process.execSync("ffmpeg -i " + Global.escapeName(filename) + " -f ffmetadata -", options);
                    var lines = (stdout + "").split("\n");
                    var amd = new AudioMetaData();
                    lines.forEach(line => {
                        if(line.startsWith("title")){
                            amd.title = line.substring("title=".length);
                        }else if(line.startsWith("artist")){
                            amd.artists = line.substring("artist=".length).split(', ');
                        }else if(line.startsWith("album_artist")){
                            amd.album_artist = line.substring("album_artist=".length);
                        }else if(line.startsWith("track")){
                            amd.track = line.substring("track=".length);
                        }else if(line.startsWith("genre")){
                            amd.genre = line.substring("genre=".length);
                        }else if(line.startsWith("date")){
                            amd.date = line.substring("date=".length);
                        }else if(line.startsWith("album")){
                            amd.album = line.substring("album=".length);
                        }
                    });
                    amd.path = filename;
                    amd.validate();
                    this.db.addSong(amd).then(() => resolve());
                }else{
                    resolve();
                }
        });
        });
    }




}

File: squirrel/Scanner.cs
using System;
using System.Text.RegularExpressions;
using System.IO;
using System.Diagnostics;
using System.Threading.Tasks;

namespace squirrel
{
	public class Scanner
	{
		Database db;
		string directory;

		public Scanner (Database db, string directory)
		{
			this.db = db;
			this.directory = directory;
		}

		public void rescan(){
			Global.putScanStatus ("Validating existing paths");
			foreach (var path in db.getAllPaths()) {
				if (!File.Exists (path)) {
					db.deletePath (path);
					Global.putScanStatus ("Deleted invalid path " + path);
				}
			}
			Global.putScanStatus ("Looking for new files");
			scanDirectory (directory);
			Global.putScanStatus ("Finished descent");
		}

		void scanDirectory(string filename){
			Global.putScanStatus ("Descending into " + filename);

			//DFS FTW
			foreach (var dir in Directory.GetDirectories(filename)) {
				scanDirectory (dir);
			}

			foreach (var file in Directory.GetFiles(filename)) {
				if (file.EndsWith (".mp3")) {
					scanFile (file);
				}
			}
		}


		void scanFile(string filename){
			if (!this.db.hasPath (filename)) {
				var psi = new ProcessStartInfo ("ffmpeg", "-i " + Global.escapeArgument (filename) + " -f ffmetadata -");
				Global.putScanStatus("Scanning " + filename);
				psi.RedirectStandardOutput = true;
				psi.RedirectStandardError = true;
				psi.UseShellExecute = false;
				var p = Process.Start (psi);
				p.WaitForExit ();

				var lines = p.StandardOutput.ReadToEnd ().Split ('\n');
				var amd = new AudioMetaData ();

				foreach (var line in lines) {
					if (line.StartsWith ("title")) {
						amd.title = line.Substring ("title=".Length);
					} else if (line.StartsWith ("artist")) {
						foreach(var artist in line.Substring ("artist=".Length).Split (',')){
							amd.artists.Add (artist);
						}
					} else if (line.StartsWith ("album_artist")) {
						amd.album_artist = line.Substring ("album_artist=".Length);
					} else if (line.StartsWith ("track")) {
						amd.track = line.Substring ("track=".Length);
					} else if (line.StartsWith ("genre")) {
						amd.genre = line.Substring ("genre=".Length);
					} else if (line.StartsWith ("date")) {
						amd.date = line.Substring ("date=".Length);
					} else if (line.StartsWith ("album")) {
						amd.album = line.Substring ("album=".Length);
					}
				}

				var e = p.StandardError.ReadToEnd ();
				var r = new Regex("\\s+Duration\\: (\\S+)\\,");
				Match m = r.Match (e);
				if (m.Success) {
					var d = m.Groups [0].ToString ().Trim();
					if (d != null && d.Length > 0) {
						try{
						var parts = d.Split (':');
						var hourpart = parts [1].Trim();
						var hours = Double.Parse(hourpart);
						var minutes = Double.Parse (parts[2].Trim());
						var seconds = Double.Parse (parts[3].Trim(' ',','));
						minutes = minutes + hours * 60;
						seconds = seconds + minutes * 60;
						amd.duration = seconds;
						}catch{
							Global.putScanStatus ("Error parsing duration in " + filename);
						}
					}
				}

				amd.path = filename;
				amd.validate ();
				this.db.addSong (amd);
			}
		}




	}
}

